### PR TITLE
Search and replace

### PR DESCRIPTION
**Pull request by kevinb7**
There are still a few things left to do:
- [x] update button state when the user can no longer replace(All)
- [x] scroll the window so that the current search result is visible
- [x] show a (current of total) indicator
- [x] remove code that handles answerArea
- [x] get ArticleEditor working
- [x] write tests
- [x] move test examples into .json files

The reason why I'm creating the pull-request right now is that I'm new to React and would like feedback on how things are structured.

The live demo has been update: http://kevinb7.github.io/perseus/

### CODE REVIEW

File: react-components [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 11eef3b87c4dc037856f3a02778724094a456c58
+Subproject commit e81b757767dd9bb5b397be6f69dc9ebe8d0dc231
```
Comment (kevinb7) [inline react-components]: I was missing a drag-and-drop component so I updated to HEAD on master.

File: src/editor-page.jsx [hunk 2]
```diff
@@ -45,8 +47,41 @@ var EditorPage = React.createClass({
...
+        if (this.state.searchIndex !== -1) {
+            var itemEditorCount = this.getItemEditorSearchCount();
+            var combinedHintsCount = this.getCombinedHintsSearchCount();
+
+            if (this.state.searchIndex < itemEditorCount) {
+                itemEditorSearchIndex = this.state.searchIndex;
+            } else if (this.state.searchIndex < itemEditorCount + combinedHintsCount) {
+                combinedHintsSearchIndex = this.state.searchIndex - itemEditorCount;
+            }
+        }
```
Comment (kevinb7) [inline src/editor-page.jsx]: This figures out what the searchIndex would be for each component if we were only considering that single component.  It could definitely use some cleanup.

File: src/editor-page.jsx [hunk 3]
```diff
@@ -88,12 +125,28 @@ var EditorPage = React.createClass({
...
+                    searchIndex={combinedHintsSearchIndex} />
+            }
+
+            {(this.props.searchAndReplace) &&
+                <SearchAndReplaceDialog
+                    ref="searchAndReplace"
+                    question={this.props.question}
+                    answerArea={this.props.answerArea}
+                    hints={this.props.hints}
+                    onChange={this.handleStateChange}
```
Comment (kevinb7) [inline src/editor-page.jsx]: There's a ```handleChange``` method but it does some stuff with props but I needed to modify the state.

File: src/item-editor.jsx [hunk 4]
```diff
@@ -31,6 +32,20 @@ var ItemEditor = React.createClass({
...
+        if (this.props.searchIndex !== -1) {
+            var questionSearchCount = Util.countOccurences(this.props.question.content, this.props.searchString);
+            var answerAreaSearchCount = Util.countOccurences(this.props.answerArea.options.content, this.props.searchString);
+            
+            if (this.props.searchIndex < questionSearchCount) {
+                questionSearchIndex = this.props.searchIndex;
+            } else if (this.props.searchIndex < questionSearchCount + answerAreaSearchCount) {
+                answerAreaSearchIndex = this.props.searchIndex - questionSearchCount;
+            }
+        }
```
Comment (kevinb7) [inline src/item-editor.jsx]: It's unfortunate having all of this search specific logic here and in hint-editor and item-editor.  Maybe I can pass down some sort of search result offsets and try to move this logic out to search-and-replace-dialog.jsx.

**Comment by kevinb7** [comment]
link to issue: https://github.com/Khan/perseus/issues/22

**Comment by ariabuckles** [comment]
Great! I'm happy to look at this but will be on planes and [I believe] without internet for most of the day; I'll get back to you as soon as I can! (with the disclaimer that a lot is going on and that might not be for a few days)

**Comment by kevinb7** [comment]
No worries.  It'll give me time to think about how to clean it up better.

File: src/editor-page.jsx [hunk 5]
```diff
@@ -160,12 +187,19 @@ var EditorPage = React.createClass({
...
         _(newProps).extend(toChange);
-        this.props.onChange(newProps, cb, silent);
+        this.props.onChange(newProps, () => {
+            if (typeof cb === "function") {
+                cb();
+            }
+            // update the search results after question, hints, or answerArea
+            // changes
+            this.refs.searchAndReplace.updateSearchResults(newProps);
+        }, silent);
```
Comment (kevinb7) [inline src/editor-page.jsx]: I decided to handle updates to questions, hints, etc. while the search panel is open.  This should make things easier both for both users (because they won't have to close the search panel to make changes) and developers (because we won't have to disable editing while the search panel is open).

File: src/editor.jsx [hunk 6]
```diff
@@ -369,7 +409,7 @@ var Editor = React.createClass({
                     var duplicate = id in widgets;
 
                     widgets[id] = this.getWidgetEditor(id, type);
-                    var classes = (duplicate || !widgets[id] ? "error " : "") +
+                    classes = (duplicate || !widgets[id] ? "error " : "") +
                             (selected ? "selected " : "");
```
Comment (kevinb7) [inline src/editor.jsx]: I removed "var" and declared "classes" at the start of the function.  This could probably be rewritten use ```cx``` to make this more React-y.

File: src/search-and-replace-dialog.jsx [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        width: 60,
+        marginRight: 8
+    },
+
+    inputStyle: {
+        display: 'inline-block',
+        float: 'left',
+        width: 220,
+        boxSizing: 'border-box'
+    },
```
Comment (kevinb7) [inline src/search-and-replace-dialog.jsx]: I'm not super happy about all the magic numbers in here.  I'm going to rewrite this using flexbox so that it's more flexible.

File: test.html [hunk 8]
```diff
@@ -166,7 +170,6 @@
             _.delay(callback, 1000, "http://fake.image.url");
         },
         apiOptions: {
-            enableOldAnswerTypes: true,
```
Comment (kevinb7) [inline test.html]: removed to make test.html look more like index.html

File: src/search-and-replace-dialog.jsx [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+    getDefaultProps: function() {
+        return {
+            onChange() {}    // TODO: throw an error
+        };
+    },
+
+    getInitialState() {
+        return {
+            replaceString: "",
```
Comment (kevinb7) [inline src/search-and-replace-dialog.jsx]: ```searchString``` and ```searchIndex``` are part of ```StatefulEditorPage```'s ```state``` because those values need to be passed down to each of the editors and their child components.  This feels like a job for a flux store, but since I haven't used flux before I'll need to do a little more investigation before I attempt this refactoring.

File: src/article-editor.jsx [hunk 10]
```diff
@@ -66,7 +67,7 @@ var ArticleEditor = React.createClass({
 
     render: function() {
 
-        return <div className="framework-perseus perseus-article">
+        return <div id="perseus" className="framework-perseus perseus-article">
```
Comment (kevinb7) [inline src/article-editor.jsx]: This adds a margin around the ArticleEditor that matches the margin around the EditorPage.

File: react-components (hunk 1 above)
Comment (ariabuckles) [inline react-components]: :+1:

File: src/editor-page.jsx (hunk 5 above)
Comment (ariabuckles) [inline src/editor-page.jsx]: Cool!

File: src/editor.jsx (hunk 6 above)
Comment (ariabuckles) [inline src/editor.jsx]: Hmmm... thanks for moving the declaration to where it's actually declared! I think the fact that we do that indicates that there are larger problems with this function though--I don't like large multiple branches setting variables--it's hard to follow.

File: test.html (hunk 8 above)
Comment (ariabuckles) [inline test.html]: Cool! I think we've removed all of the answer areas from questions live on Khan Academy, so this is a good change.

For small changes like this I support making a separate branch/commit and merging that, since it's easy to review on its own, doesn't get held up by a review of the larger change, and makes `git bisect` work better. (Though sometimes it's a lot of pain to extract fixes like these from the rest of the work, which is understandable. It's also not a bad change to make as part of a larger change; it could just probably be served better by separating it.)

File: src/article-editor.jsx (hunk 10 above)
Comment (ariabuckles) [inline src/article-editor.jsx]: Okay :). We should fix any css that relies on `#perseus` to use `.framework-perseus` (after checking that works okay where we render other uses of perseus). ids are evil!

File: src/article-editor.jsx [hunk 11]
```diff
@@ -7,11 +7,12 @@ var Editor = require("./editor.jsx");
 var EnabledFeatures = require("./enabled-features.jsx");
 var JsonEditor = require("./json-editor.jsx");
 var Renderer = require("./renderer.jsx");
+var Util = require("./util.js");
 
 var rendererProps = React.PropTypes.shape({
     content: React.PropTypes.string,
     widgets: React.PropTypes.object,
-    images: React.PropTypes.object,
+    images: React.PropTypes.object
```
Comment (ariabuckles) [inline src/article-editor.jsx]: nit: All of the browsers we support (IE9+) support trailing commas now, so we can actually leave these in. Since this is pretty recent (we dropped support for IE8 last summer) our code is pretty inconsistent, but I've been moving towards using trailing commas everywhere. They make diffs and moving lines nicer, after one gets over the weird looks.

**Comment by ariabuckles** [comment]
Okay--I've looked through most of your comments but not most of the actual code yet. I will work through that soon!

File: src/article-editor.jsx (hunk 11 above)
Comment (kevinb7) [inline src/article-editor.jsx]: That makes sense.  All of the commas should be back in place.  It'd be nice to have a .jshintrc or something like that.

File: src/util.js [hunk 12]
```diff
@@ -24,6 +24,20 @@ var Util = {
         total: 0,
         message: null
     },
+    
+    countOccurences: function(contentString, searchString) {
```
Comment (ariabuckles) [inline src/util.js]: nit: occurrences has two r's

(Disclaimer: I personally do this all the time; just pointing it out since i happened to google occurrence)

**Comment by ariabuckles** [comment]
I've played around with this a bit, and this looks really awesome. It's a bit larger of a feature, so I'm probably going to have a lot of comments on the code, but I like this a lot :).

File: src/util.js [hunk 13]
```diff
@@ -24,6 +24,20 @@ var Util = {
...
+        var count = 0;
+
+        var searchRegex = new RegExp(searchString, "g");
```
Comment (ariabuckles) [inline src/util.js]: Building up a regex dynamically like this is a bit dangerous.

If we only want the user to search for raw strings, building a regex dynamically from their input requires us to sanitize it. [MDN has an escapeRegExp function to do this sanitization if we need it](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions), or I am personally partial to writing the index based string code when doing raw string operations--there's no way to mess up escaping, and it should be faster than dynamically generated regexes.

If we want the users to be able to search using regular expressions, we'll need to wrap this call in a try/catch, since `new RegExp` can throw a `SyntaxError`, for example, if you have a regexp of `[`. The search string count and replacing seem to work when searching for a regex, but the highlighting doesn't.

I think that for our users, raw find/replace without regular expressions is probably what we want, at least to start with.

File: src/util.js [hunk 14]
```diff
@@ -24,6 +24,20 @@ var Util = {
...
+        var widgetRegex = /^\[\[\u2603 (([a-z-]+) ([0-9]+))\]\]/g;
```
Comment (ariabuckles) [inline src/util.js]: This is really similar to Util.rWidgetRule above; could we declare it up there?

Also, what does the combination of `^` and `/g` do? Testing in node, this looks to me like it ignores the `/g`, which I think is probably not what we want here.

Weird things start happening if you search for widget names right now--highlighting doesn't seem to work, while replacement does. I'm guessing that might be related to this.

File: src/util.js [hunk 15]
```diff
@@ -24,6 +24,20 @@ var Util = {
...
+
+        var matches = contentString.replace(widgetRegex, "").match(searchRegex);
+        if (matches !== null) {
+            count += matches.length;
```
Comment (ariabuckles) [inline src/util.js]: instead of having a mutable variable here, can we just `return matches.length`? and `else { return 0; }`?

File: src/util.js [hunk 16]
```diff
@@ -24,6 +24,20 @@ var Util = {
...
     },
+    
+    countOccurences: function(contentString, searchString) {
+        var count = 0;
+
+        var searchRegex = new RegExp(searchString, "g");
+        var widgetRegex = /^\[\[\u2603 (([a-z-]+) ([0-9]+))\]\]/g;
+
+        var matches = contentString.replace(widgetRegex, "").match(searchRegex);
+        if (matches !== null) {
```
Comment (ariabuckles) [inline src/util.js]: nit: unless we expect `null` and `undefined` semantics to differ, I prefer `!= null` (which is false for `null` and `undefined`, but true for everything else).

File: src/article-editor.jsx [hunk 17]
```diff
@@ -188,6 +193,11 @@ var ArticleEditor = React.createClass({
                         </div>
                     </div>
                 ];
+
+                if (section.content) {
+                    searchIndex -= Util.countOccurences(section.content, searchString);
```
Comment (ariabuckles) [inline src/article-editor.jsx]: I find doing this logic inside a .map in jsx a bit hard to follow, if just for the reason that there's a lot of template code around it.

Maybe we could precompute this as a second array before the loop? such as

```javascript
var sectionSearchCount = 0;
var cumulativeSectionSearchCounts = _.map(sections, (section) => {
    sectionSearchCount += Util.countOccurences(section.content || "", searchString);
    return sectionSearchCount;
});
```

and then pass in `searchIndex - cumulativeSectionSearchCounts[i]` as the `searchIndex` prop

File: src/article-editor.jsx (hunk 10 above)
Comment (kevinb7) [inline src/article-editor.jsx]: I had a look at editor.less and there are other styles which depend on ```#perseus``` as well as other id based rules, e.g. ```#problemarea```, ```#workarea```, ```#answer_area```, etc.  Also, there's this comment above these styles:
```
// DOM-based styles that overwrite rules set in khan-exercise.css
```
I don't want to mess stuff up further down the pipeline so I'm going to revert this change.  I'll create an issue to clean up id related styles.

File: src/util.js (hunk 12 above)
Comment (kevinb7) [inline src/util.js]: Thanks.  I should probably turn spell check back on and just add programming jargon to the dictionary file.

File: src/util.js (hunk 13 above)
Comment (kevinb7) [inline src/util.js]: Good catch.  I'm going to use an index base approach as suggest.  Having the indices will come in handy when fixing ```replaceAll``` so that it doesn't replace strings inside the widget tokens.

File: src/util.js (hunk 14 above)
Comment (kevinb7) [inline src/util.js]: I missed the ```^```.  Removing it fixes the regex.  I'll move the declaration to the top with the other regexes.

File: src/util.js (hunk 15 above)
Comment (kevinb7) [inline src/util.js]: This was left over from a previous revision where I was counting occurrences in ```question.content``` and each of the hints.  It no longer makes sense.  It will be completely replaced when the changes described above land.

File: src/util.js (hunk 16 above)
Comment (kevinb7) [inline src/util.js]: not applicable anymore but I'll keep this in mind when doing these kind of comparisons.

File: src/article-editor.jsx (hunk 17 above)
Comment (kevinb7) [inline src/article-editor.jsx]: Sounds good.  Stylistic question: why not use the built-in Array.prototype.map?

File: test.html (hunk 8 above)
Comment (kevinb7) [inline test.html]: I'm going to leave this in, but in the future I'll make small unrelated changes separately.

File: src/search-and-replace-dialog.jsx [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+                    obj.forEach(replaceFunc);
+                } else {
+                    var indices = Util.getIndicesOf(obj.content, searchString);
+                    obj.content = obj.content.replace(regex, (match, offset) => {
+                        if (indices.indexOf(offset) !== -1) {
+                            if (!replaced && globalIndex === searchIndex) {
+                                replaced = true;
+                                globalIndex++;
+                                return replaceString;
+                            } else {
```
Comment (kevinb7) [inline src/search-and-replace-dialog.jsx]: I'll add some comments in here because it's a bit hairy.

File: src/util.js [hunk 19]
```diff
@@ -25,6 +27,48 @@ var Util = {
         message: null
     },
 
+    escapeRegExp: function(string) {
+        return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
+    },
+
+    getIndicesOf: function(contentString, searchString) {
```
Comment (kevinb7) [inline src/util.js]: I'll add some comments in here to so that people know that it's ignoring over widget references

File: src/__tests__/search-and-replace-test.jsx [hunk 20]
```diff
@@ -0,0 +1,423 @@
...
+describe("Search and Replace", function() {
+
+    describe("EditorPage", function() {
+        var editor;
+        var searchInput, replaceInput;
+        var previousButton, nextButton, replaceButton, replaceAllButton;
+
+        // Note: the example question initially was using expression widgets but
+        // this caused lots of issues because this widget type uses MathJax which
+        // isn't loaded in the testing environment.
```
Comment (kevinb7) [inline src/__tests__/search-and-replace-test.jsx]: This took a while to figure out.  It worked in isolation for some reason but when running the tests together it failed.

File: src/__tests__/search-and-replace-test.jsx [hunk 21]
```diff
@@ -0,0 +1,423 @@
...
+        });
+
+        it("should replace all instances of the searchString not inside a widget reference", function() {
+            React.addons.TestUtils.Simulate.change(searchInput, {target: {value: "categorizer"}});
+            React.addons.TestUtils.Simulate.change(replaceInput, {target: {value: "equation"}});
+            React.addons.TestUtils.Simulate.click(replaceAllButton);
+
+            assert.equal(editor.state.question.content, '[[☃ categorizer 1]]\n\nequation\n\n[[☃ categorizer 2]]\n\nequation');
+            assert.equal(editor.state.hints[0].content, '[[☃ categorizer 1]]\n\nequation');
+        });
```
Comment (kevinb7) [inline src/__tests__/search-and-replace-test.jsx]: The tests themselves are pretty straight forward.

**Comment by kevinb7** [comment]
This should be ready to go now.

File: src/__tests__/search-and-replace-test.jsx (hunk 20 above)
Comment (ariabuckles) [inline src/__tests__/search-and-replace-test.jsx]: Yeah, we don't stub out mathjax (I think there's a recent mathjax-node project, but I'd rather not add that if we don't have to). If things render purely in KaTeX they're usually fine, and it's possible but tricky to do mathquill based testing for the mathquill version of the expression widget. But ultimately it's worth avoiding for tests that are focused on other things :).

File: src/__tests__/search-and-replace-test.jsx [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+    describe("EditorPage", function() {
+        beforeEach(function() {
+            editorProps = $.extend(true, editorProps, itemJson);
+            editorProps.componentClass = EditorPage;
+
+            var StatefulEditorPageFactory = React.createFactory(StatefulEditorPage);
+            editor = React.render(
+                StatefulEditorPageFactory(editorProps, null),
+                document.body
+            );
```
Comment (ariabuckles) [inline src/__tests__/search-and-replace-test.jsx]: We should be able to use [React's new spread attributes syntax](http://facebook.github.io/react/docs/jsx-spread.html) here:

```javascript
    editor = React.render(
        <StatefulEditorPage {...editorProps} />,
        document.body
    );
```

File: src/__tests__/search-and-replace-test.jsx [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+        });
+
+        afterEach(function (done) {
+            React.unmountComponentAtNode(document.body);
+            document.body.innerHTML = "";
+            setTimeout(done);
+        });
+
+        it("should replace all instances of the searchString not inside a widget reference", function() {
+            React.addons.TestUtils.Simulate.change(searchInput, {target: {value: "categorizer"}});
```
Comment (ariabuckles) [inline src/__tests__/search-and-replace-test.jsx]: Could we make these fit within 80 characters? We roughly follow https://github.com/Khan/style-guides/blob/master/style/javascript.md. `make lint` enforces this, but requires installation of the khan dotfiles right now (it runs `~/Khan/devtools/khan-linter/runlint.py`, which you can find at [khan-linter](https://github.com/Khan/khan-linter) or install via [khan-dotfiles](https://github.com/Khan/khan-dotfiles).

File: src/__tests__/search-and-replace-test.jsx [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+    describe("ArticleEditor", function() {
+        beforeEach(function() {
+            editorProps = $.extend(true, editorProps, particleJson);
+            editorProps.componentClass = ArticleEditor;
+
+            var StatefulEditorPageFactory = React.createFactory(StatefulEditorPage);
+            editor = React.render(
+                StatefulEditorPageFactory(editorProps, null),
+                document.body
+            );
```
Comment (ariabuckles) [inline src/__tests__/search-and-replace-test.jsx]: (same here)

File: src/article-editor.jsx [hunk 25]
```diff
@@ -124,6 +125,15 @@ var ArticleEditor = React.createClass({
         );
 
         var sections = this._sections();
+        var searchIndex = this.props.searchIndex;
+        var searchString = this.props.searchString;
+
+        var adjustedSearchIndices = _.map(sections, section => {
```
Comment (ariabuckles) [inline src/article-editor.jsx]: nit: we usually put parens around lambda arguments, even if there is only one.

File: src/article-editor.jsx [hunk 26]
```diff
@@ -124,6 +125,15 @@ var ArticleEditor = React.createClass({
...
+            var adjustedIndex = searchIndex;
+            var content = section.content || "";
+            searchIndex -= Util.countOccurrences(content, searchString);
+            return adjustedIndex;
+        });
```
Comment (ariabuckles) [inline src/article-editor.jsx]: thanks!

File: src/editor.jsx [hunk 27]
```diff
@@ -346,16 +348,44 @@ var Editor = React.createClass({
...
             underlayPieces = [];
 
+            var searchResultIndex = 0;
+
             for (var i = 0; i < pieces.length; i++) {
                 var type = i % 2;
                 if (type === 0) {
                     // Normal text
-                    underlayPieces.push(pieces[i]);
+                    if (this.props.searchString !== "") {
```
Comment (ariabuckles) [inline src/editor.jsx]: `if (this.props.searchString)` is a bit more concise here, and is probably what we want if searchString is null or undefined (we use editors in a couple other places, notably for the radio widget).

File: src/editor.jsx [hunk 28]
```diff
@@ -346,16 +348,44 @@ var Editor = React.createClass({
...
+                        var searchRegex =
+                            new RegExp(`(${this.props.searchString})`, "g");
```
Comment (ariabuckles) [inline src/editor.jsx]: let's escape this :)

File: src/search-and-replace-dialog.jsx [hunk 29]
```diff
@@ -0,0 +1,286 @@
+/** @jsx React.DOM */
+
+var React = require('react');
+var Util = require("./util.js");
+
+var SearchAndReplaceDialog = React.createClass({
+
+    getDefaultProps: function() {
+        return {
+            onChange() {}    // TODO: throw an error
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: None of the rest of our code uses es6 methods; could we use old es5 objects/function() for now?

I'm not sure what we'll eventually do here--es6 methods might be nice, but they also break the greppability of `updateSearchString[=:]` to find the definition of something, which is pretty valuable.

File: src/search-and-replace-dialog.jsx [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+
+    updateSearchResults() {
+        var searchIndex = 0;
+        var searchResultCount = this.getSearchResultCount(this.props.searchString);
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ searchIndex });
+    },
+
+    getSearchResultCount(searchString) {
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: I'm not sure this functionality should be part of the SearchAndReplaceDialog. We're already counting occurrences in `EditorPage` and `ArticleEditor`, so we might be able to pass the total count in to this component.

File: src/search-and-replace-dialog.jsx [hunk 31]
```diff
@@ -0,0 +1,286 @@
...
+
+        if (this.props.hints) {
+            this.props.hints.forEach(hint => {
+                count += Util.countOccurrences(hint.content, searchString);
+            });
+        }
+
+        if (this.props.particle) {
+            var particle = this.props.particle;
+            if (Array.isArray(particle)) {
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: We should probably just pass in `this._sections()` from ArticleEditor here, and then we can know that it's always an array.

File: src/search-and-replace-dialog.jsx [hunk 32]
```diff
@@ -0,0 +1,286 @@
...
+                }
+            }
+        }
+
+        return count;
+    },
+
+    handleNextSearchResult() {
+        var searchIndex = this.props.searchIndex;
+        searchIndex ++;
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: nit: no space between `searchIndex` and `++`

File: src/search-and-replace-dialog.jsx [hunk 33]
```diff
@@ -0,0 +1,286 @@
...
+        }
+
+        this.props.onChange({ searchIndex });
+    },
+
+    handleReplaceAll() {
+        var replaceString = this.state.replaceString;
+        var { searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(Util.escapeRegExp(searchString), "g");
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: yay!

File: src/search-and-replace-dialog.jsx [hunk 34]
```diff
@@ -0,0 +1,286 @@
...
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
+
+        replaceFunc(question);
+        replaceFunc(hints);
+        replaceFunc(particle);
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: This is mutating props, which isn't safe. the general approach is usually to make a copy, which probably means `_.map` / `.map` instead of `.forEach`

File: src/search-and-replace-dialog.jsx [hunk 35]
```diff
@@ -0,0 +1,286 @@
...
+
+        // handle the case where the replaceString contains one or more copies
+        // of searchString
+        // TODO(kevinb7) this is quite right
+        var searchResultCount = this.getSearchResultCount(searchString);
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: I don't think this is going to work once we fix the `replaceFunc` mutating props above, since this method relies on props. We'll probably have to pass those fields to this function (either individually or as a collective object).

File: src/search-and-replace-dialog.jsx [hunk 36]
```diff
@@ -0,0 +1,286 @@
...
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ question, hints, json: particle, searchIndex: 0 });
+    },
+
+    handleReplace() {
+        var replaceString = this.state.replaceString;
+        var { searchIndex, searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(searchString, "g");
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: let's escape this too :)

File: src/search-and-replace-dialog.jsx [hunk 37]
```diff
@@ -0,0 +1,286 @@
...
+                                globalIndex++;
+                                return match;
+                            }
+                        } else {
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: this looks really similar to `handleReplaceAll`. Maybe we can factor out the common pieces?

File: src/search-and-replace-dialog.jsx [hunk 38]
```diff
@@ -0,0 +1,286 @@
...
+        } else {
+            // edge case: replaceString contains one or more instances of
+            // searchString which means the searchResultCount will either 
+            // stay the same if the number of instances is one or increase
+            // if it's more than one
+            var matches = replaceString.match(regex);
+            if (matches) {
+                searchResultCount += matches.length - 1;
+            }
+        }
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: Can we unify these into something like:

```javascript
var matches = replaceString.match(regex);
var matchCount = matches ? matches.length : 0
searchResultCount = this.state.searchResultCount + matchCount - 1;
```

A comment about what we're doing here + why would be nice too :)

File: src/search-and-replace-dialog.jsx [hunk 39]
```diff
@@ -0,0 +1,286 @@
...
+                <br />
+                <input
+                    ref="replaceInput"
+                    type="text"
+                    value={this.props.replaceString}
+                    onChange={this.updateReplaceString}
+                    style={inputStyle} />
+                <br />
+                <div>
+                    <button
```
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: we probably want `type="button"` on each of these. the default button type is `"submit"`, which is rarely what we want (and means that when embedded in another page, it can cause weird actions)

File: src/stateful-editor-page.jsx [hunk 40]
```diff
@@ -18,18 +19,31 @@ var StatefulEditorPage = React.createClass({
...
+        return <div>
+            <this.props.componentClass {...this.state} />
+            <SearchAndReplaceDialog
+                ref="searchAndReplace"
+                question={this.state.question}
+                hints={this.state.hints}
+                particle={this.state.json}
+                searchString={this.state.searchString}
+                searchIndex={this.state.searchIndex}
+                onChange={this.handleSearchChange} />
```
Comment (ariabuckles) [inline src/stateful-editor-page.jsx]: This should go in EditorPage or ArticleEditor. StatefulEditorPage is just a wrapper around those two prop-based components for rendering as a top level component, like test.html or article.html does, and isn't actually used on khanacademy.org.

It's still find to have state on EditorPage and ArticleEditor, as long as we don't expect that state to be persisted.

File: src/util.js [hunk 41]
```diff
@@ -15,6 +15,8 @@ var Util = {
 
     rWidgetParts: /^\[\[\u2603 (([a-z-]+) ([0-9]+))\]\]$/,
     rWidgetRule:  /^\[\[\u2603 (([a-z-]+) ([0-9]+))\]\]/,
+    rWidgets: /\[\[\u2603 (([a-z-]+) ([0-9]+))\]\]/g,
+    rWidgetSplit: new RegExp("(\\[\\[\u2603 [a-z-]+ [0-9]+\\]\\])", "g"),
```
Comment (ariabuckles) [inline src/util.js]: can this use a regex literal? regex literals are generally a bit easier to read (you don't need to double escape things in literals, for example), and can sometimes be faster (since they can be compiled at parse time)

File: src/util.js [hunk 42]
```diff
@@ -25,6 +27,52 @@ var Util = {
         message: null
     },
 
+    escapeRegExp: function(string) {
```
Comment (ariabuckles) [inline src/util.js]: let's add a comment documenting where this is from

File: src/util.js [hunk 43]
```diff
@@ -25,6 +27,52 @@ var Util = {
...
+     * contentString ignoring occurrence within widget references.
+     */
+    getIndicesOf: function(contentString, searchString) {
+        if (searchString === "") {
+            return [];
+        }
+
+        var indices = [];
+        var searchStringLength = searchString.length;
+        var pieces = this.split(contentString, this.rWidgetSplit);
```
Comment (ariabuckles) [inline src/util.js]: We should probably use `Util` instead of `this`. Using `this` means that other people must always call this method using the `Util` class, and can't pull it off as a raw function (for example, `_.map(myArray, Util.getIndicesOf)` wouldn't work)

File: src/util.js [hunk 44]
```diff
@@ -25,6 +27,52 @@ var Util = {
...
+                // widget reference
+                pieceOffet += pieces[i].length;
+            }
+        }
+
+        return indices;
+    },
+
+    countOccurrences: function(contentString, searchString) {
+        return this.getIndicesOf(contentString, searchString).length;
```
Comment (ariabuckles) [inline src/util.js]: `Util` here as well

File: src/search-and-replace-dialog.jsx (hunk 18 above)
Comment (ariabuckles) [inline src/search-and-replace-dialog.jsx]: We might need to change some of this to avoid mutating props, so let's see what this looks like after that, then see if we can simplify it or add some helpful comments :)

**Comment by ariabuckles** [comment]
Sorry for the spam of comments; there's a lot of code here, but hopefully by getting them all at once it's easier to keep track of revisions over time <3

File: src/__tests__/search-and-replace-test.jsx (hunk 23 above)
Comment (kevinb7) [inline src/__tests__/search-and-replace-test.jsx]: I'll fix this.  I had some troubles running ```runlint.py```.  There are node modules which it depends on that need to be installed.  If the last line of a ```.jsx``` file is empty but jshint reports and error on it an exception will occur because python doesn't include that empty string in the array produces by ```readlines```.  I'll put together a pull request for khan-linter tomorrow to fix these issues.  It's catching the long lines which is good, but it's also reporting hundreds of other violations in react-components and other sub-repos.  Is this normal?

File: src/search-and-replace-dialog.jsx (hunk 29 above)
Comment (kevinb7) [inline src/search-and-replace-dialog.jsx]: I hadn't thought about the grep-ability, makes sense.

File: src/search-and-replace-dialog.jsx (hunk 30 above)
Comment (kevinb7) [inline src/search-and-replace-dialog.jsx]: Good point.  Moving SearchAndReplaceDialog into each of the editors will make this easy to do.

File: src/util.js (hunk 43 above)
Comment (kevinb7) [inline src/util.js]: I'm going to do this from now in all my namespace objects.

File: src/__tests__/search-and-replace-test.jsx (hunk 23 above)
Comment (ariabuckles) [inline src/__tests__/search-and-replace-test.jsx]: Well it looks like this is more broked than i thought it was. There's a lint_blacklist.txt in perseus' root directory that says to ignore everything in react-components, but then react-components must have added a lint_blacklist.txt itself since I added that, which doesn't handle all the necessary cases, so we get a bunch of errors from the examples and docs folders (among others).

I've blacklisted a few things on master of perseus and react-components, which brings the number of errors down from "insane" to "very embarrassing".

Obviously we don't have the best setup for this (but we'll try not to make it too much worse).

File: src/util.js (hunk 43 above)
Comment (ariabuckles) [inline src/util.js]: :+1:

File: src/__tests__/search-and-replace-test.jsx (hunk 23 above)
Comment (kevinb7) [inline src/__tests__/search-and-replace-test.jsx]: Here's a link to a pull request for the linter: https://github.com/Khan/khan-linter/pull/5.  It fixes the most immediate issue.  I haven't looked into why the blacklist isn't working.

We may want to look at using https://github.com/STRML/JSXHint in the future.

**Comment by kevinb7** [comment]
I am going to close this review and start one using phabricator on the weekend.

**Comment by ariabuckles** [comment]
:+1:

File: src/__tests__/search-and-replace-test.jsx (hunk 23 above)
Comment (ariabuckles) [inline src/__tests__/search-and-replace-test.jsx]: I think I'm going to try to use babel+eslint for perseus in the future (though this won't affect webapp): https://medium.com/@dan_abramov/lint-like-it-s-2015-6987d44c5b48

Comment (kevinb7) [inline src/__tests__/search-and-replace-test.jsx]: Cool.  Didn't realize that eslint could handle jsx.  This is awesomesauce.  I hope that babel+eslint spreads to other projects.  I've used babel for some of my own stuff and have been quite impressed.
